std/os.zig

```diff
@@ -3509,20 +1866,566 @@ pub const MProtectError = error{
+    if (windows.is_the_target and !builtin.link_libc) {
+        const pathname_w = try windows.sliceToPrefixedFileW(pathname);
+        return realpathW(&pathname_w, out_buffer);
+    }
+    const pathname_c = try toPosixPath(pathname);
+    return realpathC(&pathname_c, out_buffer);
+}
+
+/// Same as `realpath` except `pathname` is null-terminated.
+pub fn realpathC(pathname: [*]const u8, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+    if (windows.is_the_target and !builtin.link_libc) {
+        const pathname_w = try windows.cStrToPrefixedFileW(pathname);
+        return realpathW(&pathname_w, out_buffer);
+    }
+    if (linux.is_the_target and !builtin.link_libc) {
+        const fd = try openC(pathname, O_PATH | O_NONBLOCK | O_CLOEXEC, 0);
+        defer close(fd);
+
+        var procfs_buf: ["/proc/self/fd/-2147483648\x00".len]u8 = undefined;
+        const proc_path = std.fmt.bufPrint(procfs_buf[0..], "/proc/self/fd/{}\x00", fd) catch unreachable;
+
+        return readlinkC(proc_path.ptr, out_buffer);
+    }
+    const result_path = std.c.realpath(pathname, out_buffer) orelse switch (std.c._errno().*) {
+        EINVAL => unreachable,
+        EBADF => unreachable,
+        EFAULT => unreachable,
+        EACCES => return error.AccessDenied,
+        ENOENT => return error.FileNotFound,
+        ENOTSUP => return error.NotSupported,
+        ENOTDIR => return error.NotDir,
+        ENAMETOOLONG => return error.NameTooLong,
+        ELOOP => return error.SymLinkLoop,
+        EIO => return error.InputOutput,
+        else => |err| return unexpectedErrno(@intCast(usize, err)),
+    };
+    return mem.toSlice(u8, result_path);
+}
+
+/// Same as `realpath` except `pathname` is null-terminated and UTF16LE-encoded.
+pub fn realpathW(pathname: [*]const u16, out_buffer: *[MAX_PATH_BYTES]u8) RealPathError![]u8 {
+    const h_file = try windows.CreateFileW(
+        pathname,
+        windows.GENERIC_READ,
+        windows.FILE_SHARE_READ,
+        windows.OPEN_EXISTING,
+        windows.FILE_ATTRIBUTE_NORMAL,
+    );
+    defer windows.CloseHandle(h_file);
+
+    var wide_buf: [windows.PATH_MAX_WIDE]u16 = undefined;
+    const wide_len = try windows.GetFinalPathNameByHandleW(h_file, &wide_buf, wide_buf.len, windows.VOLUME_NAME_DOS);
+    assert(wide_len <= wide_buf.len);
+    const wide_slice = wide_len[0..wide_len];
+
+    // Windows returns \\?\ prepended to the path.
+    // We strip it to make this function consistent across platforms.
+    const prefix = []u16{ '\\', '\\', '?', '\\' };
+    const start_index = if (mem.startsWith(u16, wide_slice, prefix)) prefix.len else 0;
+
+    // Trust that Windows gives us valid UTF-16LE.
+    const end_index = std.unicode.utf16leToUtf8(out_buffer, wide_slice[start_index..]) catch unreachable;
+    return out_buffer[0..end_index];
+}
+
+/// Spurious wakeups are possible and no precision of timing is guaranteed.
+pub fn nanosleep(seconds: u64, nanoseconds: u64) void {
```

Comment:
move to `std.time`?

Reply:
`std.time` and `std.crypto` should use these routines. these are the low-level versions.

Reply:
I understand your general preference @daurnimator to move stuff out of std/os.zig. I'd like to handle that proposal separately. I think there is something to it, but I also think that the premise of this PR as it stands is sound. The comment at the top says:

```zig
// This file contains thin wrappers around OS-specific APIs, with these
// specific goals in mind:
// * Convert "errno"-style error codes into Zig errors.
// * When null-terminated byte buffers are required, provide APIs which accept
//   slices as well as APIs which accept null-terminated byte buffers. Same goes
//   for UTF-16LE encoding.
// * Where operating systems share APIs, e.g. POSIX, these thin wrappers provide
//   cross platform abstracting.
// * When there exists a corresponding libc function and linking libc, the libc
//   implementation is used. Exceptions are made for known buggy areas of libc.
//   On Linux libc can be side-stepped by using `std.os.linux` directly.
// * For Windows, this file represents the API that libc would provide for
//   Windows. For thin wrappers around Windows-specific APIs, see `std.os.windows`.
// Note: The Zig standard library does not support POSIX thread cancellation, and
// in general EINTR is handled by trying again.
```

In particular 3 of these points apply to `nanosleep`:
 * nanosleep is a libc function, available on linux, freebsd, macos, netbsd, and probably more. os.nanosleep provides cross platform abstraction over these.
 * on linux when not linking libc, os.nanosleep provides abstraction over calling the libc version or the linux syscall directly.
 * the possible error of nanosleep is converted to `void` meaning that it cannot fail (by adding "spurious wakeups are possible" to the documented behavior)

On the other hand, the decision to make it work for Windows is questionable. `nanosleep` is not a Windows API function, either directly or in msvcrt. `os.nanosleep` could have Windows support removed and that logic can go directly in std.time.sleep, and become simpler. That's probably a good idea.

Reply:
Lets continue in #2571